_posts/2023-04-14-kafka-transactions.md

```diff
@@ -0,0 +1,304 @@
+kafka.network:type=RequestMetrics,name=ResponseQueueTimeMs,request=([\w]+)
+kafka.network:type=RequestMetrics,name=ResponseSendTimeMs,request=([\w]+)
+attributes:Count,99thPercentile
+```
+
+Producer metrics are more difficult to collect, but they help when broker logs reveal some misbehavior.
+
+```sh
+kafka.producer:type=producer-metrics,client-id=([-.\w]+)
+attributes:txn-init-time-ns-total,txn-begin-time-ns-total,txn-send-offsets-time-ns-total,txn-commit-time-ns-total,txn-abort-time-ns-total
+```
+
+### Drawbacks
+
+There are a few drawbacks when using Kafka transactions that developers should be aware of.
+
+Firstly, transactions can add some latency to an application due to the write amplification required to ensure EOS.
+While this may not be a significant issue for some applications, those that require real-time processing may be affected.
+
+For each transaction we have the following overhead, which is independent of the number of messages: 
+
+- some RPCs to register the partitions with the coordinator (can be batched)
+- one marker has to be written to each partition when completing a transaction
+- few state change records are appended to the internal transaction state log
+
+Most of this latency is on the producer side, where the transactions API is implemented.
+The performance degradation can be significant when having short transaction commit intervals.
+Increasing the transaction duration also increases the end-to-end latency, so it's a matter of tradeoff.
+
+Another potential drawback is hanging transactions, which are transactions with a missing or out of order control record.
+They may be caused by a network issue causing delayed messages (see [KIP-890](https://cwiki.apache.org/confluence/display/KAFKA/KIP-890%3A+Transactions+Server-Side+Defense)).
+
+A hanging transaction is usually revealed by a stuck application with growing lag.
+Despite messages being produced, consumers can't make any progress because the LSO does not grow anymore.
+
+```sh
+[Consumer clientId=my-client, groupId=my-group] The following partitions still have unstable offsets which are not cleared on the broker side: [__consumer_offsets-27], 
+this could be either transactional offsets waiting for completion, or normal offsets waiting for replication after appending to local log
+```
+
+Additionally, if the partition belongs to a compacted topic, this causes the unbounded partition growth, because the `LogCleaner` does not clean beyond the LSO.
+If not detected and fixed in time, this may lead to disk space exhaustion and service disruption.
+
+Fortunately, there are embedded tools that can be used to find all hanging transactions and rollback them.
+After that, the LSO starts to increment again on every transaction completion and the application should be able to recover.
+
+```sh
+# find all hanging transactions in a broker
+$ kafka-transactions.sh --bootstrap-server :9092 find-hanging --broker 0
```

Comment:
Maybe put `./bin/` in front

Reply:
I think I will use `$KAFKA_HOME/bin/` everywhere, as it doesn't require to mess with the reader's PATH.